## California Housing Price Prediction

This is the start of makimg the basic housing price prediction model using real life data taken from california's housing plan. As per the reference from HOML-3

Lets begin...

In [ ]:
import subprocess
import sys

library_imports = [
    "numpy",
    "matplotlib",
    "pandas",
    "scikit-learn"
]

def install_package(package):
    subprocess.run([sys.executable, "-m", "pip", "install", package])

def freeze_packages():
    with open("requirements.txt", "w") as req_file:
        subprocess.run([sys.executable, "-m", "pip", "freeze"], stdout=req_file)

for package in library_imports:
    install_package(package)

freeze_packages()

In [ ]:
from pathlib import Path
import pandas as pd
import tarfile
import urllib.request

def load_housing_data():
    tarball_path = Path("datasets/housing.tgz")
    if not tarball_path.is_file():
        Path("datasets").mkdir(parents=True, exist_ok=True)
        url = "https://github.com/ageron/data/raw/main/housing.tgz"
        urllib.request.urlretrieve(url,tarball_path)
        with tarfile.open(tarball_path) as housing_dataset:
            housing_dataset.extractall(path="datasets")
    return pd.read_csv(Path("datasets/housing/housing.csv"))

housing = load_housing_data()

Generating housing info..

In [ ]:
housing.info()

In [ ]:
housing.isnull().sum()

In [ ]:
import matplotlib.pyplot as plt

IMAGES_PATH = Path() / "images" / "end_to_end_project"
IMAGES_PATH.mkdir(parents=True, exist_ok=True)

def save_fig(fig_id, tight_layout=True, fig_extension="png", resolution=300):
    path = IMAGES_PATH / f"{fig_id}.{fig_extension}"
    if tight_layout:
        plt.tight_layout()
    plt.savefig(path, format=fig_extension, dpi=resolution)


In [ ]:
housing.hist(bins=50,figsize=(12,8))
save_fig("9X9 Histogram Chart")
plt.show()

## Create a Test Set

In [ ]:
housing

In [ ]:
import numpy as np

def shufle_and_split(data,test_ratio):
    shuffled_indices = np.random.permutation(len(data))

    print(shuffled_indices)
    test_set_size = int(len(data)*test_ratio)
    test_indices = shuffled_indices[:test_set_size]
    train_indices = shuffled_indices[test_set_size:]

    return data.iloc[train_indices],data. iloc[test_indices]

# housing = pd.DataFrame([1,2,3,4,5,6])
train,test = shufle_and_split(housing,.2)

print("Total length: ",len(housing))
print("The Length of train:",len(train))
print("The Length of test:",len(test))

In [ ]:
# from zlib import crc32

# def is_id_in_test_set(val,set_ratio):
#     return crc32(np.int64(val)) < set_ratio * 2**32

# def splits_data_with_id(data,test_ratio,id_column):
#     ids = data[id_column]
#     in_test_set = ids.apply(lambda id_:is_id_in_test_set(id_,test_ratio))
#     return data.loc[~in_test_set],data.loc[in_test_set]

# xy = housing.reset_index()
# train,test = splits_data_with_id(xy,0.2,"index")

In [ ]:
from sklearn.model_selection import train_test_split

train,test = train_test_split(housing,test_size=0.2,random_state=42)

# Stratified Data Splitting

This process is used to split data based on their categories

In [ ]:
housing

In [ ]:
housing["income_cat"] = pd.cut(housing["median_income"],
                               bins=[0,1.5,3,4.5,6,np.inf],
                               labels=[1,2,3,4,5]
                               )

housing["income_cat"].value_counts().sort_index().plot.bar(rot=0,grid=True)
plt.xlabel("Income Category")
plt.ylabel("Number of districts")
save_fig("Income categories")
plt.show()

Time for Stratified Shuffle Split based on income data

In [ ]:
from sklearn.model_selection import StratifiedShuffleSplit

splitter = StratifiedShuffleSplit(n_splits=10,test_size=0.2,random_state=42)
strat_splits=[]
for train_index,test_index in splitter.split(housing,housing["income_cat"]):
    strat_train_split_n = housing.iloc[train_index]
    strat_test_split_n = housing.iloc[test_index]
    strat_splits.append([strat_train_split_n,strat_test_split_n])

but the above solution will give you 10 unique split of the same dataset, we can go for first element(split) [0]

or we can do it by one go

In [ ]:
strat_train_set,strat_test_set = train_test_split(
    housing,
    test_size=0.2,
    random_state=42,
    stratify=housing["income_cat"])

In [ ]:
income_prop = strat_test_set["income_cat"].value_counts()/len(strat_test_set)
income_prop

In [ ]:
#extra codes -  for computing the data

def income_cat_proportions(data):
    return data["income_cat"].value_counts()/len(data)

train_set,test_set = train_test_split(housing,test_size=0.2,random_state=42)

compare_props = pd.DataFrame({
    "Overall%":income_cat_proportions(housing),
    "Stratified%":income_cat_proportions(strat_test_set),
    "Random%":income_cat_proportions(test_set),
}).sort_index()

compare_props.index.name = "Income Category"
compare_props["Strat. Error%"] = (compare_props["Stratified%"]/compare_props["Overall%"]-1)
compare_props["Rand. Error%"] = (compare_props["Random%"]/compare_props["Overall%"]-1)

(compare_props * 100).round(2)

Now, I should drop "income_cat" column since I wont be using it again and for short time, it gave us our purpose

In [ ]:
for set_ in (strat_train_set,strat_test_set):
    set_.drop("income_cat",axis=1,inplace=True)

# Discover and Visualize the Data to Gain Insights

In [ ]:
housing = strat_train_set.copy()

In [ ]:
housing.plot(
    kind="scatter",
    x="longitude",
    y="latitude",
    grid="True",
    s=housing["population"]/100,
    label="Population",
    c="median_house_value",
    cmap="jet",
    figsize=(10,7)
    )
save_fig("The spread file of california")
plt.show()

Now, we can do is perform standard correlations between every pair of attribute using the corr() method

In [ ]:
corr_matrix = housing.corr(numeric_only=True)
corr_matrix["median_house_value"].sort_values(ascending=False)

Now lets go for visual correlatons (instead of going through entire 11 attributes, we are going for essential one)

In [ ]:
from pandas.plotting import scatter_matrix

attributes = [
    "median_house_value","median_income","total_rooms","housing_median_age"
    ]
scatter_matrix(housing[attributes], figsize=(12,8))
save_fig("Visual correlations")
plt.show()

we see that the plot in median_house_value x median_income looks promising since the plot is shown to be diagonal(close coefficient of correlation is **0.69** approx)

In [ ]:
housing.plot(kind="scatter",x="median_income",y="median_house_value",alpha=0.1,grid=True)
save_fig("Median Income x Median House Value")
plt.show()

From here, what we are going o do is attribute combination, this process is for making deep analytics on the data, like having another attribute for, lets say, number of rooms per house, the bedroom ratio and number of people that are living per house etc.

The reason for this to check out which data shows promising correlations with eachother beyond the previous data

In [ ]:
housing["rooms_per_house"] = housing["total_rooms"] / housing["households"]
housing["bedrooms_ratio"] = housing["total_bedrooms"] / housing["total_rooms"]
housing["people_per_house"] = housing["population"] / housing["households"]

In [ ]:
corr_matrix = housing.corr(numeric_only=True)
corr_matrix["median_house_value"].sort_values(ascending=False)

# Preparing data for Machine Learning algorithm



Let's revert to the original training set and separate the target (note that strat_train_set.drop() creates a copy of strat_train_set without the column, it doesn't actually modify strat_train_set itself, unless you pass inplace=True):


In [ ]:
housing = strat_train_set.drop("median_house_value", axis=1)
housing_labels = strat_train_set["median_house_value"].copy()

In [ ]:
from sklearn.impute import SimpleImputer

imputer = SimpleImputer(strategy="median")

Before we can proceed with it, we need to make a copy of the numerical data of the train set, i.e. **housing**

In [ ]:
housing_num = housing.select_dtypes(include=[np.number])

now we clean up the data

In [ ]:
imputer.fit(housing_num)

In [ ]:
X = imputer.transform(housing_num)

From here, we go with handling texts and categorical attributes.

In [ ]:
#First of all, we should look at the categorical data
housing_cat = housing[["ocean_proximity"]]
housing_cat.head(10)

We could have gone for ordinal encoder since its a categorical features but we are not about to use it, do you know why? let me show you

In [ ]:
from sklearn.preprocessing import OrdinalEncoder

ordinal_encoder = OrdinalEncoder()
housing_cat_encoded = ordinal_encoder.fit_transform(housing_cat)

In [ ]:
housing_cat_encoded[:10]

These are the category labels that are made and now, look at this..

In [ ]:
ordinal_encoder.categories_

Now, when you check this dataset and compare it with already encoded datas, you will see that the categories do not line up, NEAR BAY and  <1H OCEAN is encoded to 3 and 0 respectively, where as INLAND and NEAR OCEAN is encoded to 1 and 4, to make it simplify:

```[OCEAN,INLAND,ISLAND,<1H OCEAN,NEAR OCEAN] = [0,1,2,3,4]```

The problem with ML algorithms is that it will assume two nearby values are more similar than two distant values, for some cases, it can be fine, for example: good,better,best and excellent, but in this case, its different, it can be clearly seen seen that 0 and 4 are more similar than 0,1; thus we need to use another type of encoder.

In [ ]:
from sklearn.preprocessing import OneHotEncoder

cat_encoder = OneHotEncoder()
housing_cat_1hot = cat_encoder.fit_transform(housing_cat)

In [ ]:
housing.head(4)

# Feature Scaling and Transformation

In [ ]:
# using Min Max Scaler
from sklearn.preprocessing import MinMaxScaler

min_max_scaler = MinMaxScaler()
housing_num_min_max_scaler = min_max_scaler.fit_transform(housing_num)

In [ ]:
# using Standard Scaler
from sklearn.preprocessing import StandardScaler

standard_scaler = StandardScaler()
housing_standard_scaler = standard_scaler.fit_transform(housing_num)

In [ ]:
fig, axs = plt.subplots(1, 2, figsize=(8, 3))
housing["population"].hist(ax=axs[0],bins=50)
housing["population"].apply(np.log).hist(ax=axs[1],bins=50)
axs[0].set_xlabel("Population")
axs[1].set_xlabel("Log of population")
axs[0].set_ylabel("Number of districts")
save_fig("long_tail_plot")

In [ ]:
housing_labels.to_frame()

### Understanding the importance of `inverse_transformation()`

In [ ]:
from sklearn.linear_model import LinearRegression

target_scaler = StandardScaler()

# Using StandardScaler transformation
scaled_labels = target_scaler.fit_transform(housing_labels.to_frame())

# getting the instance of LinearRegression
model = LinearRegression()

# Applying fit estimation
model.fit(housing[["median_income"]],scaled_labels)

# Generating new data
some_new_data = housing[["median_income"]].iloc[:5]

# Training the model
scaled_predictions = model.predict(some_new_data)

# Inverse transforming the scaled predictons
prediction = target_scaler.inverse_transform(scaled_predictions)

### Why we use `inverse_transform()` after prediction

When training the model, the target values (`housing_labels`) were scaled using `StandardScaler`, producing `scaled_labels`. The LinearRegression model learns a relationship in this scaled target space. Therefore, its predictions (`scaled_predictions`) are also returned in the scaled form.

To convert these predictions back to the original units of the target variable, we apply:

```python
prediction = target_scaler.inverse_transform(scaled_predictions)



---

### **Key takeaway**
✔ `inverse_transform()` reverses scaling using the scaler’s saved parameters  
✔ The model always works in whatever scale it was trained on  

---


In order to avoid manually scaling, theres a better version to that.

### Using `TransformedTargetRegressor` with `StandardScaler`

When we want to scale the target variable `y` (for example with `StandardScaler`), we can use `TransformedTargetRegressor` to automate both the forward transform during training and the inverse transform during prediction.




In [ ]:
from sklearn.compose import TransformedTargetRegressor

model = TransformedTargetRegressor(LinearRegression(),
                                   transformer=StandardScaler())

model.fit(housing[["median_income"]],housing_labels)
predictions = model.predict(some_new_data)

## Custom Transformers

In [ ]:
from sklearn.preprocessing import FunctionTransformer

log_transformer = FunctionTransformer(np.log,inverse_func=np.exp)
log_pop = log_transformer.transform(housing[["population"]])

Silimarly, we can do the same with RBF Kernel

In [ ]:
from sklearn.metrics.pairwise import rbf_kernel


rbf_transformer = FunctionTransformer(rbf_kernel,kw_args=dict(Y=[[35]],gamma=.1))

age_simi_35 = rbf_transformer.transform(housing[["housing_median_age"]])

Now, a Trainable transformer, that can learn some parameters in the `fit()` method an using them later in `transform()` method

In [ ]:
from sklearn.base import BaseEstimator,TransformerMixin
from sklearn.cluster import KMeans

class ClusterSimilarity(BaseEstimator,TransformerMixin):
    def __init__(self,n_clusters,gamma,random_state):
        self.n_clusters = n_clusters
        self.random_state =random_state
        self.gamma = gamma

    def fit(self,X,y=None,sample_weight=None):
        self.kmeans_=KMeans(self.n_clusters,random_state = self.random_state)
        self.kmeans_.fit(X,sample_weight=sample_weight)
        return self

    def transform(self,X):
        return rbf_kernel(X, self.kmeans_.cluster_centers_,gamma=self.gamma)

    def get_feature_names_out(self,names=None):
        return [f"Cluster {i} Similarity" for i in range(self.n_clusters)]

Now we should see where it is implemented on

In [ ]:
cluster_simil = ClusterSimilarity(n_clusters=10,gamma=1.,random_state=42)
similarities = cluster_simil.fit_transform(housing[["latitude","longitude"]], sample_weight=housing_labels)

In [ ]:
similarities[:3].round(2)

In [ ]:
# extra code – this cell generates Figure 2–19

housing_renamed = housing.rename(columns={
    "latitude": "Latitude", "longitude": "Longitude",
    "population": "Population",
    "median_house_value": "Median house value (ᴜsᴅ)"})
housing_renamed["Max cluster similarity"] = similarities.max(axis=1)

housing_renamed.plot(kind="scatter", x="Longitude", y="Latitude", grid=True,
                     s=housing_renamed["Population"] / 100, label="Population",
                     c="Max cluster similarity",
                     cmap="jet", colorbar=True,
                     legend=True, sharex=False, figsize=(10, 7))
plt.plot(cluster_simil.kmeans_.cluster_centers_[:, 1],
         cluster_simil.kmeans_.cluster_centers_[:, 0],
         linestyle="", color="black", marker="X", markersize=20,
         label="Cluster centers")
plt.legend(loc="upper right")
save_fig("district_cluster_plot")
plt.show()



## Transformation Pipeline

Now let's build a pipeline to preprocess the numerical attributes:

In [ ]:
import sklearn
sklearn.set_config(display="diagram")
from sklearn.pipeline import Pipeline

num_pipeline = Pipeline([
    ("impute", SimpleImputer(strategy="median")),
    ('standardize', StandardScaler())
])

In case we dont want to name the transformers, we can just use `make_pipeline`

In [ ]:
from sklearn.pipeline import make_pipeline

num_pipeline = make_pipeline(SimpleImputer(strategy="median"), StandardScaler())

In [ ]:
housing_num_prepared = num_pipeline.fit_transform(housing_num)
housing_num_prepared[:2].round(2)

In [ ]:
df_housing_num_prepared = pd.DataFrame(housing_num_prepared, columns=num_pipeline.get_feature_names_out(),index=housing_num.index)
df_housing_num_prepared.head(3)

In [ ]:
num_pipeline.named_steps

Now lets transform both numerical and categorical pipeline together

In [ ]:
from sklearn.compose import ColumnTransformer

num_attribs = ["longitude","latitude","housing_median_age","total_rooms","total_bedrooms","population","households","median_income"]
cat_attribs = ["ocean_proximity"]

# just for reference
# num_pipeline = make_pipeline(SimpleImputer(strategy="median"), StandardScaler())

cat_pipeline = make_pipeline(
    SimpleImputer(strategy='most_frequent'),
    OneHotEncoder(handle_unknown='ignore')
)

preprocessing1 = ColumnTransformer([
    ("num",num_pipeline,num_attribs),
    ("cat",cat_pipeline,cat_attribs)
])


lengthy process since I have to include column names separately

In [ ]:

from sklearn.compose import make_column_selector,make_column_transformer

preprocessing = make_column_transformer(
    (num_pipeline, make_column_selector(dtype_include=np.number)),
    (cat_pipeline, make_column_selector(dtype_include=object))
)

# Now we are ready to apply this ColumnTransformer to the housing data
housing_prepared = preprocessing.fit_transform(housing)

In [ ]:
housing_prepared_dataframe = pd.DataFrame(housing_prepared,columns=preprocessing.get_feature_names_out(), index=housing.index)
housing_prepared_dataframe.head(3)

Lets build the pipeline fr

In [ ]:
def column_ratio(X):
    return X[:,[0]]/X[:,[1]]

def ratio_name(function_transformer, feature_names_in):
    return ["ratio"] #feature names out

def ratio_pipeline():
    return make_pipeline(SimpleImputer(strategy="median"),
                         FunctionTransformer(column_ratio,feature_names_out=ratio_name),
                         StandardScaler()
                         )

log_pipeline = make_pipeline(
    SimpleImputer(strategy="median"),
    FunctionTransformer(np.log, feature_names_out='one-to-one'),
    StandardScaler()
)
cluster_simil = ClusterSimilarity(n_clusters=10,gamma=1.,random_state=42)
default_num_pipeline = make_pipeline(SimpleImputer(strategy="median"),StandardScaler())

preprocessing = ColumnTransformer([
     ("bedrooms", ratio_pipeline(), ["total_bedrooms", "total_rooms"]),
        ("rooms_per_house", ratio_pipeline(), ["total_rooms", "households"]),
        ("people_per_house", ratio_pipeline(), ["population", "households"]),
        ("log", log_pipeline, ["total_bedrooms", "total_rooms", "population",
                               "households", "median_income"]),
        ("geo", cluster_simil, ["latitude", "longitude"]),
        ("cat", cat_pipeline, make_column_selector(dtype_include=object)),
    ],
    remainder=default_num_pipeline) # one column remaining: housing_median_age

In [ ]:
housing_prepared = preprocessing.fit_transform(housing)
housing_prepared.shape

In [ ]:
#get the features name
preprocessing.get_feature_names_out()

## Select and Train Model

In [ ]:
lin_reg = make_pipeline(preprocessing, LinearRegression())
lin_reg.fit(housing,housing_labels)

Done, now I have a working linear regression model. Now I try it out on the training set, looking at the first five predictions and comparing them to labels.

In [ ]:
housing_predictions = lin_reg.predict(housing)
housing_predictions[:5].round(2)

In [ ]:
housing_labels.iloc[:5].values

In [ ]:
# extra code – computes the error ratios discussed in the book
error_ratios = housing_predictions[:5].round(-2) / housing_labels.iloc[:5].values - 1
print(", ".join([f"{100 * ratio:.1f}%" for ratio in error_ratios]))

Well, it works, but not always: the first prediction is way off (by over
$200,000!), while the other predictions are better: two are off by about 25%,
and two are off by less than 10%. Remember that you chose to use the RMSE
as your performance measure, so you want to measure this regression model’s
RMSE on the whole training set using Scikit-Learn’s
``mean_squared_error()`` function, with the squared argument set to
``False``:

In [ ]:
try:
    from sklearn.metrics import root_mean_squared_error
except ImportError:
    from sklearn.metrics import mean_squared_error

    def root_mean_squared_error(labels, predictions):
        return mean_squared_error(labels, predictions, squared=False)

lin_rmse = root_mean_squared_error(housing_labels, housing_predictions)
lin_rmse

Since, the value is not that bad, we cant go with this, because the error value is quite large, compared to where most districts’
``median_housing_values`` range between $120,000 and $265,000, so a
typical prediction error of $68,628 is really not very satisfying. This is an
example of a model **underfitting** the training data.

In [ ]:
lin_reg.fit(housing, housing_labels)
housing_predictions = lin_reg.predict(housing)
lin_rmse = root_mean_squared_error(housing_labels, housing_predictions)
lin_rmse


So next, we try with ``DecisionTreeRegressor()``

In [ ]:
from sklearn.tree import DecisionTreeRegressor

tree_reg = make_pipeline(preprocessing,DecisionTreeRegressor(random_state=42))
tree_reg.fit(housing,housing_labels)

# After training the model, the model is then evaluated
housing_predictions = tree_reg.predict(housing)
tree_rmse = root_mean_squared_error(housing_labels,housing_predictions)
tree_rmse

It is much more likely that the model has badly overfit the data

In [ ]:
from sklearn.model_selection import cross_val_score

tree_rmses = -cross_val_score(tree_reg,
                            housing,housing_labels,
                            scoring="neg_root_mean_squared_error",
                            cv=10)

In [ ]:
pd.Series(tree_rmses).describe()

Now,the cross validation doesnt look good as it did earlier, our final straw is `RandomForestRegressor`

In [68]:
from sklearn.ensemble import RandomForestRegressor

forest_reg = make_pipeline(preprocessing,RandomForestRegressor(random_state=42))
forest_rmses = -cross_val_score(forest_reg,
                            housing,housing_labels,
                            scoring="neg_root_mean_squared_error",
                            cv=10)

In [69]:
pd.Series(forest_rmses).describe()

,0
count,10.000000
mean,47038.092799
std,1021.491757
min,45495.976649
25%,46510.418013
50%,47118.719249
75%,47480.519175
max,49140.832210


Let's compare this RMSE measured using cross-validation (the "validation error") with the RMSE measured on the training set (the "training error"):

In [70]:
forest_reg.fit(housing, housing_labels)
housing_predictions = forest_reg.predict(housing)
forest_rmse = root_mean_squared_error(housing_labels, housing_predictions)
forest_rmse


17551.2122500877

Decision Tree validation looks somewhat okay, but the training error exposes the overfitting.

## Fine Tuning the Model

This process involves tweaking the hyperparameters so that we can squeeze out better performance, since RandomForestRegressor shows a little better,compared to Decision Tree or Linear Regressor, instead of snooping around for other hyperparameters, we should go with what we have now.

Starting with, **Grid Search**

In [71]:
from sklearn.model_selection import GridSearchCV

full_pipeline = Pipeline([
    ("preprocessing", preprocessing),
    ("random_forest",RandomForestRegressor(random_state=42))
])
param_grid = [
    {
        'preprocessing__geo__n_clusters':[5,8,10],
        'random_forest__max_features':[4,6,8],
    },
    {
        'preprocessing__geo__n_clusters':[10,15],
        'random_forest__max_features':[6,8,10],
    }
]
grid_search = GridSearchCV(full_pipeline,param_grid,cv=3,scoring='neg_root_mean_squared_error')
grid_search.fit(housing,housing_labels)

GridSearchCV(cv=3,
             estimator=Pipeline(steps=[('preprocessing',
                                        ColumnTransformer(remainder=Pipeline(steps=[('simpleimputer',
                                                                                     SimpleImputer(strategy='median')),
                                                                                    ('standardscaler',
                                                                                     StandardScaler())]),
                                                          transformers=[('bedrooms',
                                                                         Pipeline(steps=[('simpleimputer',
                                                                                          SimpleImputer(strategy='median')),
                                                                                         ('functiontransformer',
                                                                                          FunctionTransformer(feature_names_out=<f...
                                                                         <sklearn.compose._column_transformer.make_column_selector object at 0x7a2dfeed1670>)])),
                                       ('random_forest',
                                        RandomForestRegressor(random_state=42))]),
             param_grid=[{'preprocessing__geo__n_clusters': [5, 8, 10],
                          'random_forest__max_features': [4, 6, 8]},
                         {'preprocessing__geo__n_clusters': [10, 15],
                          'random_forest__max_features': [6, 8, 10]}],
             scoring='neg_root_mean_squared_error')

now, we can access the best params

In [72]:
grid_search.best_params_

{'preprocessing__geo__n_clusters': 15, 'random_forest__max_features': 6}

now I can access the best estimator too

In [73]:
grid_search.best_estimator_

Pipeline(steps=[('preprocessing',
                 ColumnTransformer(remainder=Pipeline(steps=[('simpleimputer',
                                                              SimpleImputer(strategy='median')),
                                                             ('standardscaler',
                                                              StandardScaler())]),
                                   transformers=[('bedrooms',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(strategy='median')),
                                                                  ('functiontransformer',
                                                                   FunctionTransformer(feature_names_out=<function ratio_name at 0x7a2dfee...
                                                                    n_clusters=15,
                                                                    random_state=42),
                                                  ['latitude', 'longitude']),
                                                 ('cat',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  ('onehotencoder',
                                                                   OneHotEncoder(handle_unknown='ignore'))]),
                                                  <sklearn.compose._column_transformer.make_column_selector object at 0x7a2dfedd0140>)])),
                ('random_forest',
                 RandomForestRegressor(max_features=6, random_state=42))])

Finally, we can get the evaluation scores by this method

In [74]:
cv_res = pd.DataFrame(grid_search.cv_results_)
cv_res.sort_values('mean_test_score',ascending=False, inplace=True)
[...]

[Ellipsis]

In [75]:
cv_res.head()

,mean_fit_time,std_fit_time,mean_score_time,std_score_time,param_preprocessing__geo__n_clusters,param_random_forest__max_features,params,split0_test_score,split1_test_score,split2_test_score,mean_test_score,std_test_score,rank_test_score
12,10.088686,0.164570,0.176202,0.000738,15,6,"{'preprocessing__geo__n_clusters': 15, 'random...",-42725.423800,-43708.197434,-44334.935606,-43589.518946,662.417543,1
13,13.087373,0.030655,0.177320,0.000624,15,8,"{'preprocessing__geo__n_clusters': 15, 'random...",-43486.175916,-43819.842374,-44899.968680,-44068.662323,603.399271,2
6,6.601669,0.357669,0.178367,0.004284,10,4,"{'preprocessing__geo__n_clusters': 10, 'random...",-43797.854175,-44036.240246,-44960.694004,-44264.929475,501.513170,3
9,9.596264,0.359723,0.174859,0.001737,10,6,"{'preprocessing__geo__n_clusters': 10, 'random...",-43709.661050,-44163.463178,-44966.539107,-44279.887778,519.680433,4
7,9.829839,0.018109,0.174111,0.000098,10,6,"{'preprocessing__geo__n_clusters': 10, 'random...",-43709.661050,-44163.463178,-44966.539107,-44279.887778,519.680433,4


We got a number `43589.518946` which is quite better compared to `47038.092799` in RMSE of Random Forest, now we can also use another alterative, **Randomize Search**.

In [ ]:
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import randint

param_distribs = {
    'preprocessing__geo__n_clusters':randint(low=3,high=50),
    'random_forest__max_features':randint(low=2,high=20)
}
rnd_search = RandomizedSearchCV(full_pipeline,
                                param_distributions=param_distribs,
                                n_iter=10,cv=3,scoring='neg_root_mean_squared_error',
                                random_state=42
                                )

rnd_search.fit(housing,housing_labels)